sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java

```diff
@@ -238,8 +246,18 @@ public abstract void finalize(Iterable<WriteT> writerResults, PipelineOptions op
      * <p>The unique id that is given to open should be used to ensure that the writer's output does
      * not interfere with the output of other Writers, as a bundle may be executed many times for
      * fault tolerance. See {@link Sink} for more information about bundle ids.
+     *
+     * <p></p>The window and paneInfo arguments are populated when windowed writes are requested.
+     * shard and numbShards are populated for the case of static sharding. In cases where the
+     * runner is dynamically picking sharding, shard and numShards might both be set to -1.
      */
-    public abstract void open(String uId) throws Exception;
+    public abstract void open(String uId,
```

Comment:
I was expecting to see `writer.openGlobal(base, shardnum, numshards)` and `writer.openWindowed(base, window, pane, shardnum, numshards)`. This way, a writer either supports windowing or not based on the method, not nulls.

Incidentally, why does `open` not just take an already-formatted filename?